### ISSUE TITLE
HaProxy 2.7.3: SegFault in conn_backend_get

### ISSUE THREAD

**Issue by phihos** [haproxy/haproxy#2057]
### Detailed Description of the Problem

Over the weekend an L7 LB segfaulted while not being under particularly high load (~6k rps).

Opening the coredump revealed:

```
Core was generated by `/usr/sbin/haproxy -sf 1478235 -x sockpair@4 -Ws -f /etc/haproxy/haproxy.cfg -p'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000056443512acbd in conn_backend_get (s=0x7f7c65081950, srv=0x56443730a080, is_safe=<optimized out>, hash=31130838616905249) at src/backend.c:1231
1231	src/backend.c: No such file or directory.
```
 
 [This](https://github.com/haproxy/haproxy/blob/83540ed429ae527650432a501728f4ed5d377f4b/src/backend.c#L1231) is the line in question:
 
 ```c
 if (conn->mux->takeover && conn->mux->takeover(conn, i) == 0) { // ...
 ```
 
 I could not find any special conditions (e.g. mem, cpu, disk usage) that correlate with this issue.
 
 Edit: Two identically L7 LB crashed at different times (2023-02-26 15:56:08 UTC, 2023-02-26 23:24:12 UTC). They run active active and receive each  50% of the traffic. Both segfaulted at the same line in `backend.c`. If you want me to include the backtrace of the second LB just give the word.

### Expected Behavior

SegFaults should not occur.

### Steps to Reproduce the Behavior

Not sure how to reliably reproduce this.

### What is your configuration?

Config is large and I would have to strip a lot of confidentials. I can provide details if necessary.

### Output of `haproxy -vv`

```plain
HAProxy version 2.7.3-1ppa1~jammy 2023/02/15 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.3.html
Running on: Linux 5.15.0-58-generic #64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL +OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=128).
Built with OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
Running on OpenSSL version : OpenSSL 3.0.2 15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.39 2021-10-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[  OT] opentracing
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

[See gist](https://gist.github.com/phihos/f538c56283089353091d3a62545a920d)

**Comment by wtarreau** [comment]
Hi @phihos,

could you please issue "p *conn" in gdb ? I'm suspecting the mux is null and I don't see how this could happen, but regardless I'd rather make sure this is the case before trying to imagine impossible scenarios.

**Comment by phihos** [comment]
Of course, here you go:

```gdb
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000056443512acbd in conn_backend_get (s=0x7f7c65081950, srv=0x56443730a080, is_safe=<optimized out>, hash=31130838616905249) at src/backend.c:1231
1231	src/backend.c: No such file or directory.
[Current thread is 1 (Thread 0x7f7e3d168640 (LWP 1588312))]
(gdb) p *conn
$1 = {obj_type = OBJ_TYPE_CONN, err_code = 0 '\000', send_proxy_ofs = 0, flags = 525313, ctrl = 0x5644352bed20 <proto_tcpv4>, xprt = 0x0, mux = 0x0, xprt_ctx = 0x0, ctx = 0x0, 
  owner = 0x0, target = 0x56443730a080, subs = 0x0, toremove_list = {next = 0x7f7c65423c68, prev = 0x7f7c65423c68}, {session_list = {n = 0x7f7c65423c78, p = 0x7f7c65423c78}, 
    stopping_list = {n = 0x7f7c65423c78, p = 0x7f7c65423c78}}, handle = {qc = 0xfddeadfd, fd = -35738115}, proxy_netns = 0x0, destroy_cb = 0x0, src = 0x0, dst = 0x7f7c64d0dfb0, 
  proxy_authority = {ptr = 0x0, len = 0}, proxy_unique_id = {ptr = 0x0, len = 0}, hash_node = 0x7f7cad0a3530}
```

~~I am not very familiar with GDB. Maybe  I interpreted the GDB output wrong and the segfault happened somewhere else? I put the backtrace into a gist (see orginal post).~~

Edit: Ok now I understand what your command does. It dereferences the `conn` pointer and we can see that `mux` indeed does point to zero causing the segfault.

**Comment by wtarreau** [comment]
Thanks! Yeah indeed it's zero. The transport as well by the way. I have no idea how it could be stored there, we'll have to dig :-(

**Comment by phihos** [comment]
Ok, just tell me if you need any more context. Good hunting!

**Comment by capflam** [comment]
I think I have an idea of what is going on. Looking at your backtrace, we try to release the idle connection `0x7f7c65423c20` on the thread 118 because it was moved in the `toremove_conns` list. It is the list of idle connections that must be released. The release itself is handled by a dedicated task. However, on the thread 1, we extract the same connection from `safe_conns` list to reuse it. 

In the function `srv_cleanup_connections()`, idle connections are moved from `safe_conns` list or `idle_conns` list to `toremove_conns` list. However, the connection flags are not cleared. Thus, if the mux is woken up while the connection is released, it may try to reinsert it in `safe_conns` list or `idle_conns` list. In fact, when the mux's I/O function is called, the mux starts by removing the connection from any idle connection list depending on the connection flags and reinserts it before leaving.

Here is a quick and dirty patch that may solve the issue. I must review the code more carefully to be sure it it the right fix.
```diff
diff --git a/src/server.c b/src/server.c
index 6e9d5113f..ff2d27e65 100644
--- a/src/server.c
+++ b/src/server.c
@@ -5829,6 +5829,7 @@ static int srv_migrate_conns_to_remove(struct eb_root *idle_tree, struct mt_list
                hash_node = ebmb_entry(node, struct conn_hash_node, node);
                eb_delete(node);
                MT_LIST_APPEND(toremove_list, &hash_node->conn->toremove_list);
+               hash_node->conn->flags &= ~CO_FL_LIST_MASK;
                i++;
 
                node = next;
```

**Comment by phihos** [comment]
> I think I have an idea of what is going on. Looking at your backtrace, we try to release the idle connection 0x7f7c65423c20 on the thread 118 because it was moved in the toremove_conns list. It is the list of idle connections that must be released. The release itself is handled by a dedicated task. However, on the thread 1, we extract the same connection from safe_conns list to reuse it.

Would you like me to post the other backtrace to verify that assumption?

**Comment by capflam** [comment]
Yes, it can be good to double-check. Thanks

**Comment by phihos** [comment]
[Here you go](https://gist.github.com/phihos/438c82a4f8fe2b265d63ab4f5de14200). I found multiple occurrences of `srv_cleanup_toremove_conns`

**Comment by capflam** [comment]
Thanks ! It is indeed the same problem. There are several calls to `srv_cleanup_toremove_conns` and it is expected because there is one task per thread. However, here again, the thread 32 is releasing the connection we try to reuse in the thread 1.

**Comment by capflam** [comment]
Here is a patch. You may test it on top of the 2.7.3.
@wtarreau, @a-denoyelle, any comment about this patch ? 

[0001-BUG-MEDIUM-connection-Clear-flags-when-a-conn-is-rem.patch.txt](https://github.com/haproxy/haproxy/files/10851819/0001-BUG-MEDIUM-connection-Clear-flags-when-a-conn-is-rem.patch.txt)

**Comment by a-denoyelle** [comment]
Looks good to me thank you.

**Comment by wtarreau** [comment]
I'm not sure how this all works but I'm seeing this:
```
diff --git a/src/server.c b/src/server.c
index 6e9d5113f..9b299e5be 100644
--- a/src/server.c
+++ b/src/server.c
@@ -5829,6 +5829,7 @@ static int srv_migrate_conns_to_remove(struct eb_root *idle_tree, struct mt_list
 		hash_node = ebmb_entry(node, struct conn_hash_node, node);
 		eb_delete(node);
 		MT_LIST_APPEND(toremove_list, &hash_node->conn->toremove_list);
+		hash_node->conn->flags &= ~CO_FL_LIST_MASK;
 		i++;
```
IF (big if since I don't know the details) the problem is related to the fact that another thread might release that connection from the toremove_list, IMHO the line should be added before appending the connection to the list, not after, otherwise there remains a tiny race window where another thread could remove all this before the flag is removed, possibly causing a use-after-free here. But I could be wrong since I don't know that part, I'm just basing my reasoning on the code I'm seeing here involving a shared list.

**Comment by capflam** [comment]
It is performed under the per-thread idle_conns lock. So it is not possible for another thread to manipulate the idle connection list of the thread, include the mux itself. And in fact, at this place, I try to fix the opposite. We want to release a connection and we must be sure it will never be reused in the meantime.

However, I agree with the fact it is cleaner to clear the connection flags before `MT_LIST_APPEND`.

**Comment by wtarreau** [comment]
Hmmm the idle_conns are protected by the idle_conns_lock, but not the toremove_conns from what I'm seeing (and that's the purpose of having an mt_list in the first place). At least I'm not seeing here anything preventing thread 1 from calling srv_migrate_conns_to_remove() on the thread 2's toremove_conns list while thread 2 is currently running srv_cleanup_to_remove_conns(), the MT_LIST_APPEND() being done and the connection being instantly deleted by thread 2.

**Comment by capflam** [comment]
This is indeed true. This will work because when the connection is released, it also try to remove itself from idle connection list. So it tries to get the local idle_conns lock. The release will therefore be blocked. But I must admit  it's a bit far-fetched :) I will update the patch.

**Comment by capflam** [comment]
Now pushed. I switch it to fixed for now because I'm an optimistic !

**Comment by capflam** [comment]
I don't see how to backport the fix to 2.2. The same element is used to add a connection in a safe list or in the `toremove` list. I prefer to keep 2.2 safe, especially because I didn't test if it is affected by this issue. I'm closing keeping the `.2.2` label to remember it was not backported.
